File: prstk_research/engine/metrics.py
from __future__ import annotations

import math
from statistics import mean


def _returns(nav: list[float]) -> list[float]:
    return [nav[i] / nav[i - 1] - 1.0 for i in range(1, len(nav)) if nav[i - 1] > 0]


def _quantile(values: list[float], probability: float) -> float | None:
    if not values:
        return None
    ordered = sorted(values)
    index = (len(ordered) - 1) * probability
    lower, upper = math.floor(index), math.ceil(index)
    if lower == upper:
        return ordered[lower]
    weight = index - lower
    return ordered[lower] * (1 - weight) + ordered[upper] * weight


def drawdown_events(dates: list[str], nav: list[float]) -> list[dict]:
    events: list[dict] = []
    peak_index = 0
    trough_index = 0
    peak = nav[0] if nav else 0.0
    trough = peak
    for index, value in enumerate(nav):
        if value >= peak:
            if trough < peak:
                events.append({
                    "peak": dates[peak_index], "trough": dates[trough_index],
                    "recovery": dates[index], "drawdown": trough / peak - 1,
                    "peak_to_trough_days": trough_index - peak_index,
                    "trough_to_recovery_days": index - trough_index,
                    "underwater_days": index - peak_index,
                    "recovered": True,
                })
            peak, peak_index, trough, trough_index = value, index, value, index
        elif value < trough:
            trough, trough_index = value, index
    if nav and trough < peak:
        events.append({
            "peak": dates[peak_index], "trough": dates[trough_index],
            "recovery": None, "drawdown": trough / peak - 1,
            "peak_to_trough_days": trough_index - peak_index,
            "trough_to_recovery_days": None,
            "underwater_days": len(nav) - 1 - peak_index,
            "recovered": False,
        })
    return sorted(events, key=lambda item: item["drawdown"])


def calculate_metrics(rows: list[dict], *, benchmark_rows: list[dict] | None = None,
                      risk_free_rate: float = 0.0, trading_days_per_year: int = 252,
                      annual_cost: float = 0.0) -> dict:
    """Canonical metric definition for the Python research pipeline.

    ``risk_free_rate`` is annualized and is independent from borrowing or
    pledge rates.  Cost is applied to every daily NAV observation so risk
    metrics use the same net path as the reported return.
    """
    valid = [row for row in rows if math.isfinite(float(row["nav"])) and float(row["nav"]) > 0]
    if len(valid) < 2:
        return {"status": "unavailable", "reason": "insufficient_positive_nav"}
    dates = [row["date"] for row in valid]
    gross_nav = [float(row["nav"]) for row in valid]
    daily_cost = max(0.0, annual_cost) / trading_days_per_year
    net_nav = [gross_nav[0]]
    for index in range(1, len(gross_nav)):
        net_nav.append(net_nav[-1] * gross_nav[index] / gross_nav[index - 1] * max(0.0, 1 - daily_cost))
    returns = _returns(net_nav)
    years = (len(net_nav) - 1) / trading_days_per_year
    average = mean(returns) if returns else 0.0
    variance = sum((value - average) ** 2 for value in returns) / max(1, len(returns) - 1)
    volatility = math.sqrt(variance) * math.sqrt(trading_days_per_year)
    daily_rf = risk_free_rate / trading_days_per_year
    downside = [min(0.0, value - daily_rf) for value in returns]
    downside_deviation = math.sqrt(sum(value * value for value in downside) / max(1, len(downside))) * math.sqrt(trading_days_per_year)
    growth = net_nav[-1] / net_nav[0]
    gross_growth = gross_nav[-1] / gross_nav[0]
    cagr = growth ** (1 / max(years, 1 / trading_days_per_year)) - 1
    peak = net_nav[0]
    max_dd = 0.0
    for value in net_nav:
        peak = max(peak, value)
        max_dd = min(max_dd, value / peak - 1)
    events = drawdown_events(dates, net_nav)
    ordered = sorted(returns)
    var95 = _quantile(ordered, 0.05)
    var99 = _quantile(ordered, 0.01)
    cvar95_values = [value for value in returns if var95 is not None and value <= var95]
    cvar99_values = [value for value in returns if var99 is not None and value <= var99]
    result = {
        "status": "available", "start": dates[0], "end": dates[-1], "observations": len(net_nav),
        "years": years, "gross_total_return": gross_growth - 1, "total_return": growth - 1,
        "roi": growth - 1, "cagr": cagr, "annualized_return": cagr,
        "annualized_volatility": volatility, "downside_deviation": downside_deviation,
        "sharpe": (average * trading_days_per_year - risk_free_rate) / volatility if volatility else None,
        "sortino": (average * trading_days_per_year - risk_free_rate) / downside_deviation if downside_deviation else None,
        "calmar": cagr / abs(max_dd) if max_dd else None, "max_drawdown": max_dd,
        "best_day": max(returns), "worst_day": min(returns),
        "positive_day_ratio": sum(value > 0 for value in returns) / len(returns),
        "var95": var95, "cvar95": mean(cvar95_values) if cvar95_values else None,
        "var99": var99, "cvar99": mean(cvar99_values) if cvar99_values else None,
        "cost_drag": gross_growth - growth, "ending_wealth": growth, "drawdown_events": events,
        "max_drawdown_duration": events[0]["underwater_days"] if events else 0,
        "recovery_duration": events[0]["trough_to_recovery_days"] if events else None,
        "risk_free_rate": risk_free_rate,
    }
    return result

File: site/js/core/metrics.js
const finite = value => Number.isFinite(Number(value));

function validRows(rows, valueField = 'nav') {
  return (rows || [])
    .map(row => ({ ...row, nav: Number(row[valueField]) }))
    .filter(row => row.date && finite(row.nav) && row.nav > 0)
    .sort((a, b) => String(a.date).localeCompare(String(b.date)));
}

function quantile(values, probability) {
  if (!values.length) return null;
  const ordered = [...values].sort((a, b) => a - b);
  const position = (ordered.length - 1) * probability;
  const lower = Math.floor(position);
  const upper = Math.ceil(position);
  if (lower === upper) return ordered[lower];
  const weight = position - lower;
  return ordered[lower] * (1 - weight) + ordered[upper] * weight;
}

function dateDifference(start, end) {
  const a = Date.parse(`${start}T00:00:00Z`);
  const b = Date.parse(`${end}T00:00:00Z`);
  return Number.isFinite(a) && Number.isFinite(b) ? Math.round((b - a) / 86400000) : null;
}

export function returnsFromNav(rows, { valueField = 'nav' } = {}) {
  const series = validRows(rows, valueField);
  return series.slice(1).map((row, index) => ({
    date: row.date,
    value: row.nav / series[index].nav - 1,
  })).filter(row => finite(row.value));
}

export function maxDrawdown(rows, { valueField = 'nav' } = {}) {
  const series = validRows(rows, valueField);
  let peak = -Infinity;
  let peakDate = null;
  let worst = 0;
  let trough = null;
  const points = [];
  for (const row of series) {
    if (row.nav > peak) {
      peak = row.nav;
      peakDate = row.date;
    }
    const drawdown = row.nav / peak - 1;
    if (drawdown < worst) {
      worst = drawdown;
      trough = row.date;
    }
    points.push({ date: row.date, drawdown });
  }
  return { value: worst, start: peakDate, trough, points };
}

export function drawdownEpisodes(rows, { valueField = 'nav' } = {}) {
  const series = validRows(rows, valueField);
  const episodes = [];
  let peakIndex = null;
  let troughIndex = null;
  let peakNav = -Infinity;
  let minimum = 0;

  const closeEpisode = recoveryIndex => {
    if (peakIndex === null || troughIndex === null || minimum >= 0) return;
    const peak = series[peakIndex];
    const trough = series[troughIndex];
    const recovery = recoveryIndex === null ? null : series[recoveryIndex];
    episodes.push({
      peak: peak.date,
      start: peak.date,
      trough: trough.date,
      recovery: recovery?.date || null,
      drawdown: minimum,
      peakToTroughDays: troughIndex - peakIndex,
      troughToRecoveryDays: recovery ? recoveryIndex - troughIndex : null,
      underwaterDays: recovery ? recoveryIndex - peakIndex : series.length - 1 - peakIndex,
      peakToTroughCalendarDays: dateDifference(peak.date, trough.date),
      troughToRecoveryCalendarDays: recovery ? dateDifference(trough.date, recovery.date) : null,
      underwaterCalendarDays: dateDifference(peak.date, recovery?.date || series.at(-1)?.date),
      recovered: Boolean(recovery),
      recoveryIndex,
    });
  };

  series.forEach((row, index) => {
    if (row.nav >= peakNav) {
      closeEpisode(index);
      peakNav = row.nav;
      peakIndex = index;
      troughIndex = null;
      minimum = 0;
      return;
    }
    const drawdown = row.nav / peakNav - 1;
    if (drawdown < minimum) {
      minimum = drawdown;
      troughIndex = index;
    }
  });
  closeEpisode(null);
  return episodes.sort((a, b) => a.drawdown - b.drawdown);
}

function pairedReturns(assetRows, benchmarkRows, valueField = 'nav') {
  const asset = returnsFromNav(assetRows, { valueField });
  const benchmark = returnsFromNav(benchmarkRows, { valueField });
  const benchmarkByDate = new Map(benchmark.map(row => [row.date, row.value]));
  return asset.filter(row => benchmarkByDate.has(row.date))
    .map(row => [row.value, benchmarkByDate.get(row.date)]);
}

export function calculateMetrics(rows, {
  benchmarkRows = null,
  riskFreeRate = 0,
  tradingDaysPerYear = 252,
  annualCost = 0,
  valueField = 'nav',
  grossValueField = null,
} = {}) {
  const source = validRows(rows, valueField);
  if (source.length < 2) return { status: 'unavailable', reason: '資料不足' };

  const inferredGrossField = grossValueField || ((rows || []).some(row => finite(row.nav_gross)) ? 'nav_gross' : null);
  const grossSource = inferredGrossField ? validRows(rows, inferredGrossField) : source;
  const grossByDate = new Map(grossSource.map(row => [row.date, row.nav]));
  const dailyCost = Math.max(0, Number(annualCost) || 0) / tradingDaysPerYear;
  const netRows = [{ date: source[0].date, nav: source[0].nav }];
  for (let index = 1; index < source.length; index += 1) {
    const grossReturn = source[index].nav / source[index - 1].nav;
    netRows.push({
      date: source[index].date,
      nav: netRows[index - 1].nav * grossReturn * Math.max(0, 1 - dailyCost),
    });
  }

  // The recurrence above preserves an existing net path and applies an optional
  // explicit annual cost once. When no annualCost is supplied, netRows equals source.
  const returns = returnsFromNav(netRows);
  const grossGrowth = grossByDate.has(source.at(-1).date)
    ? grossByDate.get(source.at(-1).date) / grossByDate.get(source[0].date) - 1
    : source.at(-1).nav / source[0].nav - 1;
  const netGrowth = netRows.at(-1).nav / netRows[0].nav - 1;
  const years = (netRows.length - 1) / tradingDaysPerYear;
  const mean = returns.length ? returns.reduce((sum, row) => sum + row.value, 0) / returns.length : 0;
  const variance = returns.length > 1
    ? returns.reduce((sum, row) => sum + (row.value - mean) ** 2, 0) / (returns.length - 1)
    : 0;
  const volatility = Math.sqrt(variance) * Math.sqrt(tradingDaysPerYear);
  const dailyRf = riskFreeRate / tradingDaysPerYear;
  const downsideValues = returns.map(row => Math.min(0, row.value - dailyRf));
  const downsideDeviation = downsideValues.length
    ? Math.sqrt(downsideValues.reduce((sum, value) => sum + value ** 2, 0) / downsideValues.length) * Math.sqrt(tradingDaysPerYear)
    : 0;
  const cagr = Math.max(netGrowth + 1, 0) ** (1 / Math.max(years, 1 / tradingDaysPerYear)) - 1;
  const drawdown = maxDrawdown(netRows);
  const episodes = drawdownEpisodes(netRows);
  const distribution = returns.map(row => row.value);
  const var95 = quantile(distribution, 0.05);
  const var99 = quantile(distribution, 0.01);
  const cvar = threshold => {
    const tail = distribution.filter(value => value <= threshold);
    return tail.length ? tail.reduce((sum, value) => sum + value, 0) / tail.length : null;
  };
  const paired = benchmarkRows ? pairedReturns(netRows, benchmarkRows) : [];
  let beta = null;
  let alpha = null;
  let correlation = null;
  if (paired.length > 2) {
    const assetValues = paired.map(pair => pair[0]);
    const benchmarkValues = paired.map(pair => pair[1]);
    const assetMean = assetValues.reduce((sum, value) => sum + value, 0) / paired.length;
    const benchmarkMean = benchmarkValues.reduce((sum, value) => sum + value, 0) / paired.length;
    const covariance = paired.reduce((sum, pair) => sum + (pair[0] - assetMean) * (pair[1] - benchmarkMean), 0) / (paired.length - 1);
    const benchmarkVariance = benchmarkValues.reduce((sum, value) => sum + (value - benchmarkMean) ** 2, 0) / (paired.length - 1);
    const assetVariance = assetValues.reduce((sum, value) => sum + (value - assetMean) ** 2, 0) / (paired.length - 1);
    beta = benchmarkVariance ? covariance / benchmarkVariance : null;
    correlation = assetVariance && benchmarkVariance ? covariance / Math.sqrt(assetVariance * benchmarkVariance) : null;
    alpha = (assetMean - dailyRf - (beta || 0) * (benchmarkMean - dailyRf)) * tradingDaysPerYear;
  }
  return {
    status: 'available',
    start: netRows[0].date,
    end: netRows.at(-1).date,
    observations: netRows.length,
    years,
    grossTotalReturn: grossGrowth,
    totalReturn: netGrowth,
    roi: netGrowth,
    cagr,
    annualizedReturn: cagr,
    annualizedVolatility: volatility,
    downsideDeviation,
    sharpe: volatility ? (mean * tradingDaysPerYear - riskFreeRate) / volatility : null,
    sortino: downsideDeviation ? (mean * tradingDaysPerYear - riskFreeRate) / downsideDeviation : null,
    calmar: drawdown.value ? cagr / Math.abs(drawdown.value) : null,
    maxDrawdown: drawdown.value,
    maxDrawdownStart: drawdown.start,
    maxDrawdownTrough: drawdown.trough,
    maxDrawdownDuration: episodes[0]?.underwaterDays || 0,
    recoveryDuration: episodes[0]?.troughToRecoveryDays ?? null,
    bestDay: returns.length ? Math.max(...distribution) : null,
    worstDay: returns.length ? Math.min(...distribution) : null,
    positiveDayRatio: returns.length ? distribution.filter(value => value > 0).length / returns.length : null,
    var95,
    cvar95: cvar(var95),
    var99,
    cvar99: cvar(var99),
    costDrag: grossGrowth - netGrowth,
    beta,
    alpha,
    correlation,
    endingWealth: netRows.at(-1).nav / netRows[0].nav,
    riskFreeRate,
    episodes,
  };
}

export function formatMetric(value, type = 'percent') {
  if (value == null || !Number.isFinite(Number(value))) return '—';
  if (type === 'money') return new Intl.NumberFormat('zh-TW', { style: 'currency', currency: 'TWD', maximumFractionDigits: 0 }).format(value);
  if (type === 'number') return Number(value).toFixed(2);
  return `${(Number(value) * 100).toFixed(2)}%`;
}

export function monthlyReturns(rows, { valueField = 'nav' } = {}) {
  const daily = returnsFromNav(rows, { valueField });
  const buckets = new Map();
  daily.forEach(row => buckets.set(row.date.slice(0, 7), (buckets.get(row.date.slice(0, 7)) ?? 1) * (1 + row.value)));
  return [...buckets.entries()].sort(([a], [b]) => a.localeCompare(b)).map(([month, growth]) => ({ month, return: growth - 1 }));
}

export function annualReturns(rows, { valueField = 'nav' } = {}) {
  const daily = returnsFromNav(rows, { valueField });
  const buckets = new Map();
  daily.forEach(row => buckets.set(row.date.slice(0, 4), (buckets.get(row.date.slice(0, 4)) ?? 1) * (1 + row.value)));
  return [...buckets.entries()].sort(([a], [b]) => a.localeCompare(b)).map(([year, growth]) => ({ year, return: growth - 1 }));
}

export function returnDistribution(rows, { valueField = 'nav' } = {}) {
  const values = returnsFromNav(rows, { valueField }).map(row => row.value);
  const var95 = quantile(values, 0.05);
  const var99 = quantile(values, 0.01);
  const tail = threshold => {
    const valuesInTail = values.filter(value => value <= threshold);
    return valuesInTail.length ? valuesInTail.reduce((sum, value) => sum + value, 0) / valuesInTail.length : null;
  };
  return {
    count: values.length,
    mean: values.length ? values.reduce((sum, value) => sum + value, 0) / values.length : null,
    median: quantile(values, 0.5),
    p05: var95,
    p95: quantile(values, 0.95),
    var95,
    cvar95: tail(var95),
    var99,
    cvar99: tail(var99),
  };
}
